// stability-ai/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = __importDefault(require("lodash"));
const V1User = __importStar(require("./v1/user"));
const V1Engines = __importStar(require("./v1/engines"));
const V1Generation = __importStar(require("./v1/generation"));
const V2AlphaGeneration = __importStar(require("./v2alpha/generation"));
class StabilityAI {
    constructor(apiKey) {
        this.apiKey = apiKey;
    }
    get authHeaders() {
        return {
            Authorization: `Bearer ${this.apiKey}`,
        };
    }
    get orgAuthHeaders() {
        const headers = {};
        if (this.organiation)
            headers.Organization = this.organiation;
        if (this.clientId)
            headers['Stability-Client-ID'] = this.clientId;
        if (this.clientVersion)
            headers['Stability-Client-Version'] = this.clientVersion;
        return lodash_1.default.merge(this.authHeaders, headers);
    }
    get v1() {
        return {
            user: {
                account: () => V1User.account.bind(this)(),
                balance: () => V1User.balance.bind(this)(),
            },
            engines: {
                list: () => V1Engines.list.bind(this)(),
            },
            generation: {
                textToImage: (...args) => V1Generation.textToImage.bind(this)(...args),
                imageToImage: (...args) => V1Generation.imageToImage.bind(this)(...args),
                imageToImageUpscale: (...args) => V1Generation.imageToImageUpscale.bind(this)(...args),
                imageToImageMasking: (...args) => V1Generation.imageToImageMasking.bind(this)(...args),
            },
        };
    }
    get v2Alpha() {
        return {
            generation: {
                upscale: (...args) => V2AlphaGeneration.upscale.bind(this)(...args),
                upscaleResult: (...args) => V2AlphaGeneration.upscaleResult.bind(this)(...args),
                inpaint: (...args) => V2AlphaGeneration.inpaint.bind(this)(...args),
                imageToVideo: (...args) => V2AlphaGeneration.imageToVideo.bind(this)(...args),
                imageToVideoResult: (...args) => V2AlphaGeneration.imageToVideoResult.bind(this)(...args),
            },
        };
    }
}
exports.default = StabilityAI;
